### PR TITLE
Initial version of chat component API

### PR DESCRIPTION
**Pull request by minecrafter**
**This is by no means complete and will obviously require more work.**

This is an implementation of the JSON chat components present in Minecraft 1.7 and newer. All objects in the API are thread-safe and immutable (the legacy chat converter does not currently use immutable objects).
## Progress
- [x] Base API
- [x] Legacy converter
  - [x] Basic formatting and color
  - [x] URLs
- [ ] Unit tests
  - [x] Legacy converter
  - [ ] Parameter validation
  - [ ] JSON serialize/deserialize

### CODE REVIEW

**Comment by phase** [comment]
lgtm

File: src/main/java/org/fountainmc/api/chat/Component.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    public boolean isObfuscated() {
+        return obfuscated;
+    }
+
+    public Color getColor() {
+        return color;
+    }
+
+    public Optional<HoverEvent> getHoverEvent() {
```
Comment (Techcable) [inline src/main/java/org/fountainmc/api/chat/Component.java]: Why optional?
Why not just `@Nullable`?

Comment (minecrafter) [inline src/main/java/org/fountainmc/api/chat/Component.java]: Components might not have hover events, click events, or insertions.

I'm still working on this patch, however.

**Comment by PizzaCrust** [comment]
woooo gson :+1:

**Comment by minecrafter** [comment]
**Please feel free to review**. The only left for me to do is to add more tests along with other components (which we don't have ready to implement just yet), but I'll be putting this off for a later time.

File: src/main/java/org/fountainmc/api/chat/Component.java (hunk 1 above)
Comment (minecrafter) [inline src/main/java/org/fountainmc/api/chat/Component.java]: I guess I can clarify now.
- We're using Java 8, and `Optional<T>` works better with the pseudo-functional functionality.
- People may not want to handle `null` - `Optional<T>` covers those cases.

**Pull request merged**